src/Microsoft.Azure.EventHubs/EventHubsDiagnosticSource.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.EventHubs
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Diagnostic Source util for "Microsoft.Azure.EventHubs" source.
+    /// </summary>
+    internal static class EventHubsDiagnosticSource
+    {
+        public const string DiagnosticSourceName = "Microsoft.Azure.EventHubs";
+
+        public const string SendActivityName = DiagnosticSourceName + ".Send";
+        public const string SendActivityStartName = SendActivityName + ".Start";
+        public const string SendActivityExceptionName = SendActivityName + ".Exception";
+
+        public const string ReceiveActivityName = DiagnosticSourceName + ".Receive";
+        public const string ReceiveActivityStartName = ReceiveActivityName + ".Start";
+        public const string ReceiveActivityExceptionName = ReceiveActivityName + ".Exception";
+
+        internal static readonly DiagnosticListener DiagnosticListener = new DiagnosticListener(DiagnosticSourceName);
+
+        internal static bool IsEnabled => DiagnosticListener.IsEnabled();
+
+        internal static Activity StartSendActivity(string clientId, EventHubsConnectionStringBuilder csb, string partitionKey, IEnumerable<EventData> eventDatas, int count)
+        {
+            // skip if diagnostic source not enabled
+            if (!IsEnabled)
+            {
+                return null;
+            }
+
+            // skip if no listeners for this "Send" activity 
+            if (!DiagnosticListener.IsEnabled(SendActivityName,
+                new
+                {
+                    Endpoint = csb.Endpoint,
+                    EntityPath = csb.EntityPath,
+                    PartitionKey = partitionKey
+                }))
+            {
+                return null;
+            }
+
+            Activity activity = new Activity(SendActivityName);
+
+            // extract activity tags from input
+            activity.AddTag("component", "Microsoft.Azure.EventHubs");
```

Comment:
❤️ , 👍 for setting tags and for using the opentracing conventions!!!

Reply:
While I agree with @cwe1ss, I want to have the rationale for all of the tags used.

We know about two kinds of DiagnosticSource listeners:
1. specific to particular source - they parse payloads
2. generic - they use tags

Tags are overhead  - tracing system has to add them to the telemetry item, pass over the wire and store, so I want to make sure ALL this tags are essential.

I.e.
-  "component", "Microsoft.Azure.EventHubs: it's the same as DiagnosticSource Name. Why don;t we create a policy that DiagnosticSource name is a component
- "peer.service", "Azure Event Hub": it's the same information, as component isn't it?
- "operation.data": $"{csb.EntityPath}/{partitionKey}" and "eh.event_hub_name"/"eh.partition_key" - pure duplication
- "operation.name" is in fact ActivityName

Reply:
In a/this particular case these points are valid, however it is not a rule, e.g.:
- "component" identifies this particular SDK, but per [DS user guide](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md#diagnosticlistener-names) multiple sources can be used in a single library for different purposes
- "operation.name" at this point happens to map 1-to-1 with activity, but e.g. for DocumentDB, Redis it is many-to-1
- "peer.service" identifies the "other" side of operation - e.g. there are numerous libraries for Redis - all of them would have different DS name, but all of them are talking to the same "Redis" service

Obviously here we should focus on what makes most sense for EventHub SDK and we can agree on fall-back paths to imply different properties, however if we want to maintain generic traceability of different sources we need to have a minimal common set of tags.